cli/src/check.rs

```diff
@@ -0,0 +1,197 @@
+        CheckResult {
+            description,
+            ..Default::default()
+        }
+    }
+}
+
+pub async fn traffic_split_check(client: Client) -> CheckResult {
+    let api = Api::<CustomResourceDefinition>::all(client.clone());
+    let mut result = CheckResult::new("TrafficSplit CRD exists");
+    match api.get_opt("trafficsplits.split.smi-spec.io").await {
+        Result::Ok(Some(_)) => result.result = CheckStatus::Success,
+        Result::Ok(None) => {
+            result.result = CheckStatus::Error;
+            result.error = Some("TrafficSplit CRD is not installed".into());
+            result.hint = Some("https://github.com/linkerd/linkerd-smi");
+        }
+        Result::Err(err) => {
+            result.result = CheckStatus::Error;
+            result.error = Some(err.to_string());
+            result.hint = Some("https://github.com/linkerd/linkerd-failover#troubleshooting");
+        }
+    };
+    result
+}
+
+pub async fn namespace_check(client: Client) -> (CheckResult, Option<String>) {
+    let mut result = CheckResult::new("failover extension namespace exists");
+    let extension_label = ListParams::default().labels("linkerd.io/extension=failover");
+    let api = Api::<Namespace>::all(client.clone());
+    let ns_list = api.list(&extension_label).await;
+    match ns_list {
+        Result::Ok(ref objs) if objs.items.len() == 1 => {
+            result.result = CheckStatus::Success;
+            let ns = objs.items.first().expect("failover namespace must exist");
+            (result, Some(ns.name()))
+        }
+        Result::Ok(ref objs) if objs.items.is_empty() => {
+            result.result = CheckStatus::Error;
+            result.error = Some("Failover namespace not found".into());
+            result.hint = Some("https://github.com/linkerd/linkerd-failover#troubleshooting");
+            (result, None)
+        }
+        Result::Ok(_) => {
+            result.result = CheckStatus::Error;
+            result.error = Some("Multiple failover namespaces found".into());
+            result.hint = Some("resulthttps://github.com/linkerd/linkerd-failover#troubleshooting");
+            (result, None)
+        }
+        Result::Err(err) => {
+            result.result = CheckStatus::Error;
+            result.error = Some(err.to_string());
+            result.hint = Some("https://github.com/linkerd/linkerd-failover#troubleshooting");
+            (result, None)
+        }
+    }
+}
+
+pub async fn deploy_check(client: Client, ns: &str) -> CheckResult {
+    let mut result = CheckResult::new("failover controller is healthy");
+    let api = Api::<Deployment>::namespaced(client, ns);
+    match api.get_opt("linkerd-failover").await {
+        Result::Ok(Some(deploy)) => {
+            if deploy.status.map_or(false, |status| {
+                status
+                    .available_replicas
+                    .map_or(false, |replicas| replicas > 0)
+            }) {
```

Comment:
nit, take it or leave it: this if condition feels complex enough i might stick it in a `let`?
```suggestion
            let has_available_replicas = deploy
                .status
                .map_or(false, |status| {
                    status
                        .available_replicas
                        .map_or(false, |replicas| replicas > 0)
                });

            if has_available_replicas {
```